copy/copy.go

```diff
@@ -124,17 +125,30 @@ type ImageListSelection int
 
 // Options allows supplying non-default configuration modifying the behavior of CopyImage.
 type Options struct {
-	RemoveSignatures                 bool            // Remove any pre-existing signatures. SignBy will still add a new signature.
-	SignBy                           string          // If non-empty, asks for a signature to be added during the copy, and specifies a key ID, as accepted by signature.NewGPGSigningMechanism().SignDockerManifest(),
-	SignPassphrase                   string          // Passphare to use when signing with the key ID from `SignBy`.
-	SignBySigstorePrivateKeyFile     string          // If non-empty, asks for a signature to be added during the copy, using a sigstore private key file at the provided path.
-	SignSigstorePrivateKeyPassphrase []byte          // Passphare to use when signing with `SignBySigstorePrivateKeyFile`.
+	RemoveSignatures bool // Remove any pre-existing signatures. SignBy will still add a new signature.
+
+	// Signers to use to add signatures during the copy.
+	// Callers are still responsible for closing these Signer objects; they can be reused for multiple copy.Image operations in a row.
+	Signers []*signer.Signer
```

Comment:
I don't have a strong opinion on that.  One worry I have is boilerplate code for callers.  skopeo, libimage and CRI-O would all need updates, wouldn't they?

Reply:
Totally, this would need a common utility. (Not CRI-O, it doesn’t create signatures.)

Right now there is a WIP https://github.com/mtrmac/image/tree/fulcio-signing/pkg/cli/sigstore : just importing `copy.Image` adds dependencies enough to _validate_ Fulcio + Rekor signatures, and to sign using private keys, but importing this `cli/sigstore` subpackage drags in the Fulcio + Rekor RPC clients.

I actually fairly strongly think that we need to go further than that, and to expose a config file for signing with Fulcio+Rekor (because there are like 6 individual options to set). And then `NewSignerFromSigstoreSigningConfigFile` would be a natural utility to share across various `copy.Image` callers to implement that config file.


---

If the comment is about the deprecation of existing `SignBy…` options, I now think that’s something that is eventually a good idea, but should _not_ be done in this PR; it’s an internal cleanup for no end-user benefit and the end-user functionality is much more important short-term.

Reply:
SGTM

I am mostly worried to end up like in a pre-libimage period where the tools have duplicate code that must be maintained separately.

A config file sounds good to me and makes perfect sense.  It seems more user friendly (and code friendly) to have one central place to configure the setup rather than passing things from the CLI through the entire call stack.